lib/blockchain.go

```diff
@@ -2633,15 +2640,172 @@ func (bc *Blockchain) processBlockPoW(desoBlock *MsgDeSoBlock, verifySignatures
+	var validatorSet []*ValidatorEntry
+	// 1e. Validate QC
+	if err := bc.validateQC(desoBlock, validatorSet); err != nil {
+		return false, false, nil, err
+	}
+
+	// @sofonias @piotr - should we move this to
+	// If the block doesn’t contain a ValidatorsTimeoutAggregateQC, then that indicates that we
+	// did NOT timeout in the previous view, which means we should just check that
+	// the QC corresponds to the previous view.
+	if desoBlock.Header.ValidatorsTimeoutAggregateQC.isEmpty() {
+		// The block is safe to vote on if it is a direct child of the previous
+		// block. This means that the parent and child blocks have consecutive
+		// views. We use the current block’s QC to find the view of the parent.
+		// TODO: Any processing related to the block's vote QC.
+	} else {
+		// If we have a ValidatorsTimeoutAggregateQC set on the block, it means the nodes decided
+		// to skip a view by sending TimeoutMessages to the leader, so we process
+		// the block accordingly.
+		// 1f. If timeout QC, validate that block hash isn't too far back from the latest.
+		if err := bc.validateTimeoutQC(desoBlock, validatorSet); err != nil {
+			return false, false, nil, err
+		}
+		// TODO: Get highest timeout QC from the block.
+		// We find the QC with the highest view among the QCs contained in the
+		// AggregateQC.
+		var highestTimeoutQC *QuorumCertificate
+		// TODO: Check if our local highestQC has a smaller view than the highestTimeoutQC.
+		// If our local highestQC has a smaller view than the highestTimeoutQC,
+		// we update our local highestQC.
+		if highestTimeoutQC.ProposedInView > bc.highestQC.ProposedInView {
+			// Update our high QC.
+			// TODO: Should probably COPY this.
+			bc.highestQC = highestTimeoutQC
+		}
+	}
+
+	// 2. We can now add this block to the block index since we have performed
+	// all basic validations
+	if err := bc.addBlockToBlockIndex(desoBlock); err != nil {
+		return false, false, nil, err
+	}
+
+	// 3. Determine if we're missing a parent block of this block and any of its parents from the block index.
+	// If so, add block to block index and return the hash of the missing block.
+	missingBlockHash, err := bc.validateAncestorsExist(desoBlock)
+	if err != nil {
+		return false, false, nil, err
+	}
+	if missingBlockHash != nil {
+		return false, true, []*BlockHash{missingBlockHash}, nil
+	}
+
+	// 4. Handle reorgs if necessary
+	if bc.shouldReorg(desoBlock) {
+		if err = bc.handleReorg(desoBlock); err != nil {
+			return false, false, nil, err
+		}
+	}
+
+	// Happy path
+	// 5. Add block to best chain.
+	if err = bc.addBlockToBestChain(desoBlock); err != nil {
+		return false, false, nil, err
+	}
+
+	// 6. Update in-memory struct holding uncommitted blocks.
+	if err = bc.updateUncommittedBlocks(desoBlock); err != nil {
+		return false, false, nil, err
+	}
+
+	// 7. Commit grandparent if possible.
+	if err = bc.commitGrandparents(desoBlock); err != nil {
+		return false, false, nil, err
+	}
```

Comment:
Worth thinking about how to order these such that an intermediate failure in one of them doesn't leave the node in a weird state.

Reply:
I think the proper order should be:
1. Add the block to the best chain. If this fails for any reason, we can't commit grandparents and we can't update uncommitted block ever, so it's a good one to handle first.
2. Commit grandparents. We'll do this second as we need to have the new block on the best chain before we can run the commit rule. 
3. Updating uncommitted blocks should probably be done in two pieces as I think we need to do the following:

- Add this new block to the uncommittedBlocksMap, this is trivial
- Pruning the uncommittedBlocksMaps. This we only need to do if we hit the commit rule OR the reorg rule.
- Neither of these should return an error.